pkg/debug/subsystem.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// StatusFunc is a function returning the debug status of a subsytem. It is
+// passed into RegisterStatusFunc().
+type StatusFunc func() string
+
+// StatusMap is the collection of debug status of all subsystems. The key is
+// the subsystem name. The value is the subsystem debug status.
+type StatusMap map[string]string
+
+// StatusObject is the interface an object must impelement to be able to be
+// passed into RegisterStatusObject().
+type StatusObject interface {
+	// DebugStatus() is the equivalent of StatusFunc. It must return the
+	// debug status as a string.
+	DebugStatus() string
+}
+
+type functionMap map[string]StatusFunc
+
+type statusFunctions struct {
+	functions functionMap
+	mutex     lock.RWMutex
```

Comment:
Does it need the mutex? From what I'm seeing in this PR, the register is only called in a single place and it's during the initialization.